lib/handler/file.c

```diff
@@ -110,6 +119,46 @@ static void do_proceed(h2o_generator_t *_self, h2o_req_t *req)
         do_close(&self->super, req);
 }
 
+static void do_multirange_proceed(h2o_generator_t *_self, h2o_req_t *req) {
+    struct st_h2o_sendfile_generator_t *self = (void *)_self;
+    size_t rlen, *range_cur = self->range_infos + 2 * self->finished_range, used_buf = 0;
+    ssize_t rret;
+    h2o_iovec_t vec;
+
+    if (self->bytesleft == 28) {
+        vec.len = sprintf(self->buf, "\r\n--%s--\r\n", self->boundary.base);
+        vec.base = self->buf;
+        h2o_send(req, &vec, 1, 1);
+        do_close(&self->super, req);
+        return;
```

Comment:
For the purpose of adding the _end-of-multipart_ marker, I would suggest allocating a new buffer from memory pool to store the marker, and pass it to h2o_send() at once with the last content (i.e. the second argument to the function should contain two iovecs, first one containing the content, and the second one containing the _end-of-multipart_ marker, and the third argument should be set to _2_ instead of _1_).

By doing that way the code would become simpler, and you would no longer to adjust `bytesleft`.